Add new Puppeteer front-end e2e test: checkout page

juliaamosova opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

1. Add a new front-end e2e test: checkout page;
2. Update README with Docker setup details. 

### Changes proposed previously and now a part of the migration branch (important to know in order to test this PR):

1) New `docker-compose.yaml` and `Dockerfile` in order to run e2e tests locally using Docker;
2) New `@jest/test-sequencer` package in order to run tests in a particular order;
3) New `activate-and-setup/setup-wizard.test.js` - new e2e test that will run before all other tests;
4) New `test:e2e-dev` script for running e2e tests in `dev` mode (headless mode off and SloMo mode on);
5) New config file `test:e2e.json` to pull variables from, needed to run e2e tests.
6) Travis implementation for running e2e tests with Docker. 

More details on each point:

1) **Docker**: this PR adds Dockerfiles needed to build a test site locally using Docker. An [official WordPress Docker image](https://github.com/docker-library/docs/blob/master/wordpress/README.md#-via-docker-stack-deploy-or-docker-compose) is used to build the site.  Once the site using the WP Docker image is built, the current WooCommerce dev branch is being copied to the `plugins` folder of that newly built test site. No WooCommerce Docker image is being built or needed. 

2) **Jest test sequencer**: by default, jest runs tests ordered by the time it takes to run the test (the test that takes longer to run will be run first, the test that takes less time to run will run last). Jest sequencer introduces tools that can be used to specify the order in which the tests are being run. In our case, they are being run in alphabetical order of the directories where tests are located. This way, tests in the new directory `activate-and-setup` will run first. 

3) **Setup Wizard E2E test**: this test (located in `activate-and-setup` directory) will run before all other tests. This will allow making sure that WooCommerce is activated on the site and for the setup wizard to be completed on a brand new install of WooCommerce + it will also test Setup Wizard flow. 

4) **`test:e2e-dev` script**: this allows to run `npm run test:e2e-dev` in order to run e2e test suite in dev mode. Previously, we had to add `--dev` flag to the `test:e2e` script. 

5) **`test:e2e.json` config file**: needed so that we won't need to set test site URL, admin and customer details manually anymore. 

6) **Travis**: we can't test the Travis implementation in this PR since this branch was not made of the master. We need to merge this PR to our migration branch (https://github.com/woocommerce/woocommerce/pull/24218) which is of the master to see if the Travis implementation works and how tests run there. 

### How to test the changes in this Pull Request:

1. Install Docker Desktop if you don't have it installed:

- [Docker Desktop for Mac](https://docs.docker.com/docker-for-mac/install/)
- [Docker Desktop for Windows](https://docs.docker.com/docker-for-windows/install/)

Once installed, you should see `Docker Desktop is running` message with the green light next to it (Docker Desktop for Mac):

![docker](https://user-images.githubusercontent.com/19143190/69484051-891ac300-0e26-11ea-84ac-93a8d235173e.jpg)

**Note, that if you install docker through other methods such as homebrew, for example, your steps to set it up will be different. The commands listed in steps below may also vary.** 

2. Checkout the branch of this PR to test and stay on this branch. 

Note that by default, the local test site will be built to access on `http://localhost:8084/`. If you need to change the port (from `8084` to another one), open `docker-compose.yaml` file and change it on lines 23 and 44. 

3. Run `nmp install`

4. Run `composer install --no-dev`

5. Run `npm run build`

6. Run the following command to build the test site using Docker: `docker-compose up` and watch the site being built. Note that it may take a few minutes the first time you do that. 
7. Open new terminal window and `cd` to the current branch again.
8. Run the following command to make sure the containers were built and running: `docker ps`. You should see the 2 following containers: 

- `woocommerce_wordpress-woocomerce-dev`
- `mariadb:10.4`

9. Navigate to `http://localhost:8084/` to view the newly built site. 

10. Since we specify the site URL and create admin and customer users in docker-compose file, there is `./tests/e2e-tests/config/default.json` file that contains the variables needed to run the test:

```
{
  "url": "http://localhost:8084/",
  "users": {
    "admin": {
      "username": "admin",
      "password": "password"
    },
    "customer": {
      "username": "customer",
      "password": "password"
    }
  }
}
```

If you changed either site URL or users details in the docker-compose file, you'd need to copy the content of the `default.json`, paste it to `test:e2e.json` and edit it further there to match your own setup. 

11. We can now run the tests by running: 

- `npm run test:e2e` (headless)
- `npm run test:e2e-dev` (dev mode: non-headless with SloMo)

12. Make sure all tests are passing. 

**Note that after you run the test suite for the first time and complete the Setup Wizard**, you will need to skip the Setup Wizard next time you want to run test suite again on the same install. To skip the tests, use `.only` in the relevant test entry to specify the tests that you do want to run. 

For example, in order to skip Setup Wizard tests, add `.only` to the login and activation tests as follows in the `setup-wizard.test.js`:

```
it.only( 'Can login', async () => {}
```

```
it.only( 'Can make sure WooCommerce is activated. If not, activate it', async () => {}
```

As a result, when you run `setup-wizard.test.js`, only the login and activate tests will run. The rest will be skipped. You should see the following in the terminal:

```
 PASS  tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js (11.927s)
  Store owner can login and make sure WooCommerce is activated
    ✓ Can login (7189ms)
    ✓ Can make sure WooCommerce is activated. If not, activate it (1187ms)
  Store owner can go through store Setup Wizard
    ○ skipped Can start Setup Wizard
    ○ skipped Can fill out Store setup details
    ○ skipped Can fill out Payment details
    ○ skipped Can fill out Shipping details
    ○ skipped Can fill out Recommended details
    ○ skipped Can skip Activate Jetpack section
    ○ skipped Can finish Setup Wizard - Ready! section
  Store owner can finish initial store setup
    ○ skipped Can enable tax rates and calculations
    ○ skipped Can configure permalink settings
```

<hr>

**TO STOP & RESTART DOCKER:**

- Press `Ctrl+C` in the terminal window where the containers are running 
- Stop the container(s) using the following command: `docker-compose down`
- Restart the containers using the following command: `docker-compose up`

**TO STOP DOCKER & DO A CLEAN RESTART:**

Steps below will allow building a brand new site with a clean DB and no data as it was built initially:

- Press `Ctrl+C` in the terminal window where the containers are running 
- Stop the container(s) and delete all volumes using the following command: `docker-compose down -v`
- Restart the containers using the following command: `docker-compose up --build`

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

juliaamosova committed:
Add new Puppeteer front-end e2e test: checkout page

rodrigoprimo commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,151 @@
+		await createSimpleProduct();
+
+		// Go to general settings page
+		await StoreOwnerFlow.openSettings( 'general' );
```
I think it is fine to merge this PR with this code to adjust the settings, but IMHO it would be better to have this configuration in another place. Especially considering that everything that is configured here is pretty standard. Maybe in a place that is executed only once as I mentioned in the comment above or in a base database file. I know we already talked a little bit about this and we disagree. I'm happy to proceed as is and then we can reevaluate in the future.

juliaamosova replied:
We are now configuring initial settings in the `setup-wizard.test.js` test that runs before all other tests. However, I find it is still important to verify the settings in each test since they can be different for each test.

rodrigoprimo commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,151 @@
+		await CustomerFlow.addToCartFromShopPage( 'Simple product' );
+		await CustomerFlow.goToCheckout();
+
+		await expect( page ).toFill( '#billing_first_name', 'John' );
```
Should we abstract the code to fill the address into its own function to avoid repeating it below?

juliaamosova replied:
Good catch, thanks! I addressed that by creating appropriate functions for filling out billing and shipping addresses.

rodrigoprimo commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,151 @@
+		await uiUnblocked();
+		await CustomerFlow.placeOrder();
+
+		await expect( page ).toMatch( 'Order received' );
```
I wonder if besides just checking if the message is displayed, we should also check if the order actually exists and if its contents match what is expected? This way we could catch more issues then if we simply check for the message.

juliaamosova replied:
Thanks for this feedback! I added additional order verification by the store owner in wp-admin once the order had been created.

rodrigoprimo commented on tests/e2e-tests/utils/components.js:
```diff
@@ -0,0 +1,129 @@
+	await verifyAndPublish();
+};
+
+export { createSimpleProduct };
```
Why we are exporting here just `createSimpleProduct` and not the other two functions defined in this file? Maybe it is something that is obvious to JS developers but I don't follow as I have little experience with this language.

juliaamosova replied:
We are now also exporting `createVariableProduct()` but I think you are referring to `verifyAndPublish()`? If so, `verifyAndPublish()` is currently only used within `tests/e2e-tests/utils/components.js` which is why we don't export it.

rodrigoprimo reviewed (commented):
I left some comments but none of them are blockers in my opinion.

In my environment, not all tests are passing. This is what I get:

```
npm run test:e2e -- --dev

> woocommerce@3.7.0 test:e2e /home/rodrigo/devel/vagrant-local/www/wccore/public_html/wp-content/plugins/woocommerce
> ./tests/bin/e2e-test-integration.js "--dev"

 FAIL  tests/e2e-tests/specs/front-end/front-end-checkout.test.js (115.83s)
  Checkout page
    ✕ should displays cart items in order review (6220ms)
    ✓ allows customer to choose available payment methods (6121ms)
    ✓ allows customer to fill billing details (18627ms)
    ✓ allows customer to fill shipping details (15364ms)
    ✓ allows guest customer to place order (22848ms)

  ● Checkout page › should displays cart items in order review

    expect(received).resolves.toHaveLength(expected)

    Expected length: 1
    Received length: 0
    Received array:  []

      48 | 			']';
      49 | 
    > 50 | 		await expect( page.$x( checkoutItemXPath ) ).resolves.toHaveLength( 1 );
         | 		                                                      ^
      51 | 	},
      52 | };
      53 | 

      at Object.args [as toHaveLength] (node_modules/expect/build/index.js:202:20)
      at Object.toHaveLength [as productIsInCheckout] (tests/e2e-tests/utils/flows.js:50:57)
      at Object.productIsInCheckout (tests/e2e-tests/specs/front-end/front-end-checkout.test.js:75:22)

 FAIL  tests/e2e-tests/specs/wp-admin/wp-admin-product-new.test.js (55.475s)
  Add New Product Page
    ✓ can create simple virtual product titled "Simple Product" with regular price $9.99 (15420ms)
    ✕ can create product with variations (33405ms)

  ● Add New Product Page › can create product with variations

    TimeoutError: Element .woocommerce_variation:nth-of-type(1) .handlediv not found

      waiting for function failed: timeout 1000ms exceeded

Test Suites: 2 failed, 2 total
Tests:       2 failed, 5 passed, 7 total
Snapshots:   0 total
Time:        171.586s
Ran all test suites.
```

Not sure if there is something wrong with my environment or the tests. Are all the testing passing for you? If so, can you help me debug the tests so that I understand why they are not passing in my environment?

juliaamosova committed:
Resolve conflicts with migration branch

juliaamosova committed:
Resolving conflicts with feature branch

juliaamosova committed:
Update products in the checkout verification & address filling functions

juliaamosova committed:
Implement @jest/test-sequencer in order to run tests in specific order

juliaamosova commented:
This PR is blocked for a review until https://github.com/woocommerce/woocommerce/pull/25105 is merged to https://github.com/woocommerce/woocommerce/pull/24218.

juliaamosova committed:
Resolving conflict with migration branch

juliaamosova committed:
Add order verification by the store owner

juliaamosova committed:
Add additional order verification by the store owner

juliaamosova committed:
Update README with Docker setup details and running tests

juliaamosova commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,151 @@
+
+describe( 'Checkout page', () => {
+	beforeAll( async () => {
+		await activatePlugin( 'woocommerce' );
```
In order not to run `activatePlugin()` in every test, we are now checking if the plugin is activated only once in the `setup-wizard.test.js` test that runs before all other tests.

peterfabian reviewed (commented):
Hi, the tests run fine, all green, great work! 

Couple of nitpicks below, but pre-approving.

The only thing I don't like is that I need to edit the test file for it to run fine the second time around. As we allow the setup wizard to run from the admin GUI repeatedly, it is a valid use case to run it again. That being said, we're changing the setup wizard soon anyway, so let's not spend too much time e2e testing the existing one.

juliaamosova committed:
Resolving conflicts with migration branch

juliaamosova committed:
Various improvements to e2e tests

juliaamosova commented:
> The only thing I don't like is that I need to edit the test file for it to run fine the second time around. As we allow the setup wizard to run from the admin GUI repeatedly, it is a valid use case to run it again. That being said, we're changing the setup wizard soon anyway, so let's not spend too much time e2e testing the existing one.

Thanks for a thorough review @peterfabian! 

Regarding Setup Wizard test needing to run only once - for now, I addressed it by introducing a check for the Setup Wizard notice:

- If setup wizard notice is present on the page, run Setup Wizard test;
- If not, skip it.

That way, we won't need to edit tests when we need to run the entire test suite for the 2nd time. 

Once new OBW is in place, we can see how we can improve it by introducing Docker container / DB snapshot as you proposed before.

juliaamosova commented on tests/e2e-tests/README.md:
```diff
@@ -6,108 +6,180 @@ Automated end-to-end tests for WooCommerce.
+- Run `npm run build`
 
-To run e2e tests use the following command:
+- Run the following command to build the test site using Docker: `docker-compose up` and watch the site being built. Note that it may take a few minutes the first time you do that. The process is considered completed when the messages letting you know that WordPress was installed, WooCommerce was activated and users created will be displayed:
```
As far as I understand, when you run `docker-compose up` for the first time, it includes `--build` by default. It is only if you need to rebuild existing container, you'd include `--build`. Hence, I think we can leave it as-is.

peterfabian replied:
I see. I was thinking more from the perspective of forgetful myself coming back after some time of not running this regularly, looking through this guide. I thought I might forget to run with the `--build` switch. In other words, running using that switch all the time seems safer to me, if I want to avoid having strange problems, so I thought why not add it here as well.

juliaamosova commented on tests/e2e-tests/README.md:
```diff
@@ -126,18 +198,20 @@ The following packages are being used to write tests:
-For more Puppeteer debugging tips, check [Google's documentation](https://developers.google.com/web/tools/puppeteer/debugging).   
+- Press `Ctrl+C` in the terminal window where the containers are running 
+- Stop the container(s) and delete all volumes using the following command: `docker-compose down -v`
+- Restart the containers using the following command: `docker-compose up --build`
```
Thanks!

juliaamosova commented on tests/e2e-tests/config/default.json:
```diff
@@ -9,5 +9,46 @@
+        "state": "CA",
+        "postcode": "94107"
+      }
+    }
```
Yes, I think it is a good idea to include several profiles depending on where the store or customers are located but I would rather introduce them when we'd need them so it could be tested right away.

peterfabian replied:
Yup, was just pondering about how to add new test cases, seems like your solution supports it nicely.

juliaamosova commented on tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js:
```diff
@@ -41,12 +43,12 @@ describe( 'Store owner can go through store Setup Wizard', () => {
+		await expect( page ).toFill( '#store_postcode', config.get( 'addresses.admin.store.postcode' ) );
 
 		// Select currency and type of products to sell details
 		await expect( page ).toSelect( 'select[name="currency_code"]', '\n' +
```
Added comments for more clarity, thanks for pointing it out.

juliaamosova commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,188 @@
+		await verifyCheckboxIsSet( '#woocommerce_paypal_enabled' );
+
+		await StoreOwnerFlow.logout();
+	} );
```
I see how settings can be configured differently but not sure what do you mean by "add some kind of first step in that direction". I imagine we'd add whatever we need when we'd be writing new tests so that it could be tested right away?

peterfabian replied:
You're right, let's leave this for when we actually add further tests.

juliaamosova commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,188 @@
+			config.get( 'addresses.customer.billing.state' ),
+			config.get( 'addresses.customer.billing.postcode' ),
+			config.get( 'addresses.customer.billing.phone' ),
+			config.get( 'addresses.customer.billing.email' )
```
Good call, thanks. Fixed it as you suggested.

juliaamosova commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,188 @@
+		await expect( page ).toMatch( 'Order received' );
+
+		// Get order ID from the Thank you page URL
+		let reg = /.+?\:\/\/.+?(\/.+?)(?:#|\?|$)/;
```
I agree with you that getting order ID from HTML is more stable than from URL since it can change in the future if we decide to set different permalinks settings. Thanks for pointing it out. Fixed it as you suggested.

juliaamosova commented on tests/e2e-tests/specs/front-end/front-end-checkout.test.js:
```diff
@@ -0,0 +1,188 @@
+		await expect( page ).toMatchElement( '.woocommerce-order-data__heading', { text: 'Order #' + orderId + ' details' } );
+
+		// Verify product name
+		await expect( page ).toMatchElement( '.wc-order-item-name', { text: 'Simple product' } );
```
I agree. I added simple and variable product names in the `defaut.json` file so we could get it from there and change only there if and when needed. I edited some tests with new details including the checkout page test but will need to work a bit more to update the rest of the tests. I will do it in the final migration PR if that is OK.

juliaamosova committed:
Wait for the Recommended section to load in the setup wizard test

peterfabian commented:
To sum up, on Friday, I saw 2 new errors:

```
Peters-MacBook-Pro:woocommerce peter$ docker ps
CONTAINER ID        IMAGE                                  COMMAND                  CREATED             STATUS              PORTS                  NAMES
08deea5b30da        woocommerce_wordpress-woocomerce-dev   "docker-entrypoint.s…"   29 seconds ago      Up 12 seconds       0.0.0.0:8084->80/tcp   woocommerce_wordpress-woocomerce-dev_1
327f168dd476        mariadb:10.4                           "docker-entrypoint.s…"   30 seconds ago      Up 28 seconds       3306/tcp               woocommerce_db_1
Peters-MacBook-Pro:woocommerce peter$ npm run test:e2e

> woocommerce@3.7.0 test:e2e /Users/peter/vagrant-local/www/wordpress-one/public_html/wp-content/plugins/woocommerce
> ./tests/bin/e2e-test-integration.js

 FAIL  tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js (19.252s)
  Store owner can login and make sure WooCommerce is activated
    ✓ can login (7212ms)
    ✓ can make sure WooCommerce is activated. If not, activate it (1521ms)
  Store owner can go through store Setup Wizard
    ✕ can start Setup Wizard when visiting the site for the first time. Skip all other times. (3800ms)
  Store owner can finish initial store setup
    ✓ can enable tax rates and calculations (2463ms)
    ✓ can configure permalink settings (1877ms)

  ● Store owner can go through store Setup Wizard › can start Setup Wizard when visiting the site for the first time. Skip all other times.

    Error: failed to find element matching selector "#wc_recommended_storefront_theme"

      144 | 			// Fill out recommended section details
      145 | 			// Turn off Storefront Theme option
    > 146 | 			await page.$eval( '#wc_recommended_storefront_theme', elem => elem.click() );
          | 			^
      147 |
      148 | 			// Turn off Automated Taxes option
      149 | 			await page.$eval( '#wc_recommended_automated_taxes', elem => elem.click() );

      at ElementHandle.$eval (node_modules/puppeteer/lib/JSHandle.js:477:13)
          at runMicrotasks (<anonymous>)
      at Object.<anonymous> (tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js:146:4)
        -- ASYNC --
      at ElementHandle.<anonymous> (node_modules/puppeteer/lib/helper.js:111:15)
      at DOMWorld.$eval (node_modules/puppeteer/lib/DOMWorld.js:156:21)
          at runMicrotasks (<anonymous>)
      at Object.<anonymous> (tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js:146:4)
        -- ASYNC --
      at Frame.<anonymous> (node_modules/puppeteer/lib/helper.js:111:15)
      at Page.$eval (node_modules/puppeteer/lib/Page.js:348:29)
      at Page.<anonymous> (node_modules/puppeteer/lib/helper.js:112:23)
      at Object.$eval (tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js:146:15)
          at runMicrotasks (<anonymous>)

 PASS  tests/e2e-tests/specs/front-end/front-end-cart.test.js (21.937s)
  Cart page
    ✓ should displays no item in the cart (1997ms)
    ✓ should add the product to the cart when "Add to cart" is clicked (1860ms)
    ✓ should increase item qty when "Add to cart" of the same product is clicked (2051ms)
    ✓ should update qty when updated via qty input (1219ms)
    ✓ should remove the item from the cart when remove is clicked (757ms)
    ✓ should update subtotal in cart totals when adding product to the cart (3188ms)
    ✓ should go to the checkout page when "Proceed to Checkout" is clicked (1880ms)

 FAIL  tests/e2e-tests/specs/front-end/front-end-checkout.test.js (34.691s)
  Checkout page
    ✓ should display cart items in order review (2737ms)
    ✓ allows customer to choose available payment methods (3059ms)
    ✓ allows customer to fill billing details (2168ms)
    ✕ allows customer to fill shipping details (1769ms)
    ✓ allows guest customer to place order (3963ms)
    ✓ store owner can confirm the order was received (3872ms)

  ● Checkout page › allows customer to fill shipping details

    Node is either not visible or not an HTMLElement



 PASS  tests/e2e-tests/specs/front-end/front-end-my-account.test.js (9.52s)
  My account page
    ✓ allows customer to login (5370ms)
    ✓ allows customer to see orders (1057ms)
    ✓ allows customer to see downloads (929ms)
    ✓ allows customer to see addresses (770ms)
    ✓ allows customer to see account details (1077ms)

 PASS  tests/e2e-tests/specs/front-end/front-end-single-product.test.js (30.959s)
  Single Product Page
    ✓ should be able to add simple products to the cart (3136ms)
  Variable Product Page
    ✓ should be able to add variation products to the cart (3319ms)

 PASS  tests/e2e-tests/specs/wp-admin/wp-admin-coupon-new.test.js (5.878s)
  Add New Coupon Page
    ✓ can create new coupon (3665ms)

 PASS  tests/e2e-tests/specs/wp-admin/wp-admin-order-new.test.js (6.132s)
  Add New Order Page
    ✓ can create new order (3945ms)

 PASS  tests/e2e-tests/specs/wp-admin/wp-admin-product-new.test.js (21.864s)
  Add New Simple Product Page
    ✓ can create simple virtual product titled "Simple Product" with regular price $9.99 (4129ms)
  Add New Variable Product Page
    ✓ can create product with variations (13777ms)

 PASS  tests/e2e-tests/specs/wp-admin/wp-admin-settings-general.test.js (8s)
  WooCommerce General Settings
    ✓ can update settings (5806ms)

 PASS  tests/e2e-tests/specs/wp-admin/wp-admin-settings-product.test.js (5.272s)
  WooCommerce Products > Downloadable Products Settings
    ✓ can update settings (3102ms)

 PASS  tests/e2e-tests/specs/wp-admin/wp-admin-settings-tax.test.js (23.562s)
  WooCommerce Tax Settings
    ✓ can enable tax calculation (2091ms)
    ✓ can set tax options (1972ms)
    ✓ can add tax classes (3152ms)
    ✓ can set rate settings (2407ms)
    ✓ can remove tax classes (11744ms)

Test Suites: 2 failed, 9 passed, 11 total
Tests:       2 failed, 34 passed, 36 total
Snapshots:   0 total
Time:        187.468s
Ran all test suites.
```

I played with it a bit more today, my findings:
- It seems to me that `productIsInCheckout` does not perform a good check, I can type in any quantity or amount as total/subtotal and it passes with a green check; 
- however, I have no idea why it started failing to find the `ship-to-different-address-checkbox` element
- changing `await expect( page ).toClick( '#ship-to-different-address-checkbox' );` to

```
await page.evaluate(() => {
			document.querySelector('#ship-to-different-address-checkbox').click();
		});
```

 as per this [Issue report](https://github.com/puppeteer/puppeteer/issues/2977) helps. Not sure if it's a good solution, though.

juliaamosova committed:
Update clicks in obw and checkout tests & productIsInCheckout function

juliaamosova committed:
Wait for Recommended section elements to be visible in the obw test

juliaamosova committed:
Update the way `ship to different address` checkbox is set

juliaamosova commented:
@peterfabian thanks for additional testing and debugging. 

Regarding the error in the setup-wizard test:

>Error: failed to find element matching selector "#wc_recommended_storefront_theme"

Either the test is trying to click on the element before the page is loaded or the click doesn't get registered. 

I already added waiting for the page to load before clicking on the storefront element here:

https://github.com/woocommerce/woocommerce/pull/24676/files#diff-1346c6cce433e508ba9f45a142bdd379R145

Now I added additional verification for making sure the element is visible on the page:

`await page.waitForSelector( '#wc_recommended_storefront_theme', { visible: true } );`

Regarding the error in the checkout test:

>changing await expect( page ).toClick( '#ship-to-different-address-checkbox' ); to

>await page.evaluate(() => {
			document.querySelector('#ship-to-different-address-checkbox').click();
		});

Similar to the storefront element issue above, I added additional verification for making sure that `#ship-to-different-address-checkbox` element is visible on the page and I updated the way we select the checkbox:

```
await page.waitForSelector( '#ship-to-different-address-checkbox', { visible: true } );
await setCheckbox( '#ship-to-different-address-checkbox' );
```

> It seems to me that productIsInCheckout does not perform a good check, I can type in any quantity or amount as total/subtotal and it passes with a green check;

The issue was caused by the following:

Here is the `productIsInCheckout()` function:

```
productIsInCheckout: async ( productTitle, quantity, total, cartSubtotal ) => {
		await expect( page ).toMatchElement( '.product-name', { text: productTitle } );
		await expect( page ).toMatchElement( '.product-quantity', { text: quantity } );
		await expect( page ).toMatchElement( '.product-total .amount', { text: total } );
		await expect( page ).toMatchElement( '.cart-subtotal .amount', { text: cartSubtotal } );
	},
```

The element is being matched by text and text must be a string. 

Previously, when I was passing amounts to verify as integers which caused for verification not to work:

`await CustomerFlow.productIsInCheckout( simpleProductName, 4, 39.96, 39.96 );`

I updated the amounts to pass as strings, not integers:

`await CustomerFlow.productIsInCheckout( simpleProductName, "4", "39.96", "39.96" );`

It now works as expected. You can test by changing the amount of any amount parameter and you'd see that the test will fail. It will also tell you what was expected and what was presented instead. Thanks for catching this!

juliaamosova committed:
Merge branch 'new/puppeteer-screenshot-tester' into new/front-end-checkout-e2e-test

juliaamosova committed:
Add check for new setup wizard flow & fix ship to different address click

juliaamosova commented:
@peterfabian I made additional tweaks:

>Error: failed to find element matching selector "#wc_recommended_storefront_theme"

We found out that Storefront isn’t offered in the Recommended step for those with WP 5.3 default theme. For now, I commented out the code that looks for the "#wc_recommended_storefront_theme" selector and unchecks that box. We can add it back as soon as it is fixed. 

>changing await expect( page ).toClick( '#ship-to-different-address-checkbox' ); to

>await page.evaluate(() => {
document.querySelector('#ship-to-different-address-checkbox').click();
});

I implemented the click as per the above. I will work on understanding why the regular click doesn't work here. 

I also added a check for the new setup wizard:

- If the user is presented with the new setup wizard option, the test will click on `Continue with the old setup wizard`;

- If not, the test will continue to the old setup wizard by default.

peterfabian reviewed (approved):
All tests pass for me now, both headless and with the `-dev` switch. 

I also applied fixes from #25216 and tested with uncommenting the Storefront checking/unchecking and tests passed, so I'd say we should be safe to keep it in there.
